compiler/injectdestructors.nim

```diff
@@ -888,7 +888,11 @@ proc p(n: PNode; c: var Con; s: var Scope; mode: ProcessMode; tmpFlags = {sfSing
     of nkCallKinds:
       if n[0].kind == nkSym and n[0].sym.magic == mEnsureMove:
         inc c.inEnsureMove
-        result = p(n[1], c, s, sinkArg)
+        if isAnalysableFieldAccess(n[1], c.owner) and isLastRead(n[1], c, s) and
+            hasDestructor(n[1].typ):
+          result = p(n[1], c, s, normal)
+        else:
+          result = p(n[1], c, s, sinkArg)
         dec c.inEnsureMove
         return
```

Comment:
I think the real problem here is your special case and the `return`. Instead the usual logic should be performed where the diagnosis intercepts `c.inEnsureMove`.

Reply:
I made the return because `ensureMove` has different logic with normal calls. The bug is that the `result` should have been destroyed if it is sunk.

Reply:
Most logic of normal calls are not needed for `ensureMove`. The bug is that when a value is moved to a location (`sinkArg`), the location should own the value and ensure its destruction. (Another way is to translate `ensure(x)` into a `move(x)` call instead of `=sink(x))

Reply:
So where did you get this logic from?

Reply:
> Instead the usual logic should be performed where the diagnosis intercepts c.inEnsureMove.

Okay, I removed the special treatment. Though I need to pass the paramenter of `ensureMove` with `sinkArg`?